src/coord/src/coord.rs

```diff
@@ -5182,54 +5183,68 @@ mod timeline {
     ///
     /// At each time, writes happen and then reads happen, meaning that writes at a time are
     /// visible to exactly reads at that time or greater, and no other times.
-    enum TimestampOracleState {
+    enum TimestampOracleState<T> {
         /// The timeline is producing collection updates timestamped with the argument.
-        Writing(mz_repr::Timestamp),
+        Writing(T),
         /// The timeline is observing collections aot the time of the argument.
-        Reading(mz_repr::Timestamp),
+        Reading(T),
     }
 
     /// A type that provides write and read timestamps, reads observe exactly their preceding writes..
     ///
     /// Specifically, all read timestamps will be greater or equal to all previously reported write timestamps,
     /// and strictly less than all subsequently emitted write timestamps.
-    pub struct TimestampOracle {
-        state: TimestampOracleState,
-        advance_to: Option<mz_repr::Timestamp>,
+    pub struct TimestampOracle<T> {
+        state: TimestampOracleState<T>,
+        advance_to: Option<T>,
+        next: Box<dyn Fn() -> T>,
     }
 
-    impl TimestampOracle {
-        /// Create a new timeline, starting at the indicated time.
-        pub fn new(initially: mz_repr::Timestamp) -> Self {
+    impl<T: super::CoordTimestamp> TimestampOracle<T> {
+        /// Create a new timeline, starting at the indicated time. `next` generates
+        /// new timestamps when invoked. The timestamps have no requirements, and can
+        /// retreat from previous invocations.
+        pub fn new<F>(initially: T, next: F) -> Self
```

Comment:
can `next` be part of the `CoordTimestamp` trait?

Reply:
I don't think so, because this trait is going to be used for more than this PR. (Oh, already merged here: https://github.com/MaterializeInc/materialize/blob/82a0de1f6259a995674e3b2a25893adc585658d1/src/coord/src/coord.rs#L5376). I'm not sure how it'd have the ability to provide a next fn that might depend on external state.

Reply:
I think that is generally tricky, without having the timestamp hold on to a reference to the `now` function, which I bet we don't want.